File: Remote/Source/Remote Elements/Commands/Command Subclasses/ActivityCommand.m
//
//  ActivityCommand.m
//  Remote
//
//  Created by Jason Cardwell on 4/10/13.
//  Copyright (c) 2013 Moondeer Studios. All rights reserved.
//
#import "Command_Private.h"
#import "Activity.h"
#import "Remote-Swift.h"

static int       ddLogLevel   = LOG_LEVEL_DEBUG;
static const int msLogContext = (LOG_CONTEXT_COMMAND | LOG_CONTEXT_FILE | LOG_CONTEXT_CONSOLE);
#pragma unused(ddLogLevel, msLogContext)

@interface ActivityCommandOperation : CommandOperation @end


@implementation ActivityCommand

@dynamic activity;

+ (ActivityCommand *)commandWithActivity:(Activity *)activity {
  ActivityCommand * command = [self commandInContext:activity.managedObjectContext];
  command.activity = activity;
  return command;
}

- (MSDictionary *)JSONDictionary {
  MSDictionary * dictionary = [super JSONDictionary];

  SafeSetValueForKey(self.activity.commentedUUID, @"activity.uuid", dictionary);

  [dictionary compact];
  [dictionary compress];

  return dictionary;
}

- (void)awakeFromInsert {
  [super awakeFromInsert];

  self.indicator = YES;
}

- (void)updateWithData:(NSDictionary *)data {
  /*
     {
     "class": "activity",
     "activity.uuid": "8C1918C0-300F-4DBF-AC09-B8DB9CF55290" // Sonos Activity
     }
   */

  [super updateWithData:data];

  NSDictionary * activity = data[@"activity"];

  if (activity)
    self.activity = [Activity importObjectFromData:activity context:self.managedObjectContext];

}

- (CommandOperation *)operation { return [ActivityCommandOperation operationForCommand:self]; }

- (NSString *)shortDescription {
  return $(@"activity: %@", (self.activity.name ?: @"nil"));
}

@end

@implementation ActivityCommandOperation

- (void)main {
  @try {
    Activity * activity = ((ActivityCommand *)_command).activity;
    [activity launchOrHault:^(BOOL success, NSError * error)
    {
      _success = success;
      _finished = YES;
      [super main];
    }];
  } @catch(NSException * exception)   {
    MSLogDebugTag(@"wtf?");
  }
}

@end

File: Remote/Source/Remote Elements/Commands/Command Subclasses/ActivityCommand.swift
//
//  ActivityCommand.swift
//  Remote
//
//  Created by Jason Cardwell on 3/2/15.
//  Copyright (c) 2015 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

@objc(ActivityCommand)
class ActivityCommand: Command {

  @NSManaged var activity: Activity?

  /**
  JSONDictionary

  :returns: MSDictionary!
  */
  override func JSONDictionary() -> MSDictionary {
    let dictionary = super.JSONDictionary()

    dictionary["class"] = "activity"
    safeSetValue(activity?.commentedUUID, forKey: "activity.uuid", inDictionary: dictionary)

    dictionary.compact()
    dictionary.compress()

    return dictionary
  }

  /** awakeFromInsert */
  override func awakeFromInsert() {
    super.awakeFromInsert()
    indicator = true
  }

  /**
  updateWithData:

  :param: data [NSObject:AnyObject]!
  */
  override func updateWithData(data: [NSObject:AnyObject]!) {
    super.updateWithData(data)
    if let activityData = data["activity"] as? [NSObject:AnyObject], let moc = managedObjectContext,
      let activity = Activity.importObjectFromData(activityData, context: moc) {
        self.activity = activity
    }
  }

  override var operation: CommandOperation {
    return ActivityCommandOperation(forCommand: self)
  }
}
